lib/ex_doc/formatter/html.ex

```diff
@@ -22,27 +24,36 @@ defmodule ExDoc.Formatter.HTML do
     protocols  = filter_list(:protocols, all)
 
     has_readme = config.readme && generate_readme(output, module_nodes, config, modules, exceptions, protocols)
-    generate_overview(modules, exceptions, protocols, output, config, has_readme, config.main)
+    generate_index(output, config)
+    generate_overview(modules, exceptions, protocols, output, config, has_readme)
     generate_sidebar_items(modules, exceptions, protocols, output)
     generate_list(modules, all, output, config, has_readme)
     generate_list(exceptions, all, output, config, has_readme)
     generate_list(protocols, all, output, config, has_readme)
 
-    if config.main do
-      generate_main(output, config)
+    Path.join(config.output, "index.html")
+  end
+
+  defp build_config(config) when is_map(config) do
+    if config.main == "index" do
+      raise ArgumentError, message: "`main` cannot be set to \"index\", otherwise it will recursively link to itself"
     end
 
-    Path.join(config.output, "index.html")
+    preconfig = %{
+      main: "overview",
+    }
+    Map.merge(config, preconfig, fn(_k, v1, v2) ->
+      if is_nil(v1) do v2; else v1; end
+    end)
```

Comment:
This is equivalent to:

    Map.update(config, :main, "overview", fn v -> v || "overview" end)

Or:

    Map.put(config, :main, config[:main] || "overview")

Both shorter and clearer than the current merge approach. :)